src/libsyntax_pos/symbol.rs

```diff
@@ -143,9 +144,15 @@ impl Decodable for Ident {
     }
 }
 
-/// A symbol is an interned or gensymed string.
+/// A symbol is an interned or gensymed string. The use of newtype_index! means
+/// that Option<Symbol> only takes up 4 bytes, because newtype_index! reserves
+/// the last 256 values for tagging purposes.
 #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct Symbol(u32);
+pub struct Symbol(SymbolIndex);
+
+newtype_index! {
+    pub struct SymbolIndex { .. }
```

Comment:
Any particular reason not to make `Symbol` a `newtype_index` directly?

Reply:
Hmm, probably not!

Reply:
Actually: I tried it, and it won't work because `Symbol` has its own implementations of `fmt::Debug`, `Encodable` and `Decodable` that are different to the ones for a `newtype_index!`. I will add a comment about this.

Reply:
While it is possible to coerce the macro into allowing you to implement these manually, it's a rarely used feature and probably not worth keeping around